articles/ros2_security_contexts.md

```diff
@@ -0,0 +1,307 @@
+However, having to book keep both identifier spaces simulations may introduce to many degrees of freedom that a human could groc or easily introspect via tooling.
+
+#### `<push_ros_namespace namespace="..." context="foo"/>`
+
+TODO: Describe added `context` attribute to `push_ros_namespace` element.
+Keeps pushing contexts close/readable to pushing of namespaces.
+
+#### `<push_ros_context context="foo"/>`
+
+TODO: Describe added `push_ros_context` element.
+Keeps pushing context path independent/flexable from namespaces.
+
+
+## Concerns
+
+### Multiple namespaces per context
+
+For circumstances where users may compose multiple nodes of dissimilar namespaces into a single context, the user must subsequently specify a common fully qualified context path for each node to compose, as the varying different namespaces would not push to a common context.
+For circumstances where the context path is orthogonal to node namespace, the use of fully qualifying all relevant nodes is could be tedious, but could perhaps could still be parametrized via the use of `<var/>`, and `<arg/>` substitution and expansion.
+
+
+### Modeling permissions of nodes in a process v.s. permission of the middleware ``Participant``
+
+Before the use of contexts, multiple nodes composed into a single process where each mapped to a separate ``Participant``.
+Each ``Participant`` subsequently load a security identity and access control credential prevalent to its' respective node.
+However, all nodes in that process share the same memory space and can thus access data from other nodes.
+There is a mismatch between the middleware credentials/permissions loaded and the resources accessible within the process.
+
+By using contexts, all nodes in a context share the same security identity and access control credentials.
+This inevitably means that code compiled to node ``foo`` can access credentials/permissions only trusted to node ``bar``.
+This consequence of composition could unintendedly subvert the minimal spanning policy as architected by the policy designer or measured/generated via ROS 2 tooling/IDL.
+
+With the introduction of contexts, it becomes possible to describe the union of access control permission by defining a collection of SROS 2 policy profiles as element within a specific context.
+This would allow for formal analysis tooling [2] to check for potential violations in information flow control given the composing of nodes at runtime.
+If a process contains a single context, this reconciles the permissions of a ``Participant`` and the ones of the process.
+
+However, should multiple contexts be used per process, then such security guaranties are again lost because both contexts will share the same memory space.
+Thus it should be asked whether if multiple contexts per process should even be supported.
```

Comment:
I think that the main reason to map a `Participant` to a `Context` and not one `Participant` per process, is to allow domain bridge applications.

Reply:
For such bridge applications, what would be a layer of abstraction to model the union of contexts over different domains? I'd like to empathize this question, as bridge interfaces are particularly targeted for exploiting flaws in access control and information flow control. Think of CAN bus to infotainment systems, or LAN to AP radios. Thus retaining a level of abstraction to encapsulate this, be it contexts or something else, could go a long way in aiding formal analyses of such precarious interfaces.

Reply:
IIUC, to bridge two `Domains` in DDS you need two `DomainParticipants` (one in each domain). Each `DomainParticipant` needs their own security artifacts.
I don't understand why another level of abstraction would be needed to encapsulate this.

Reply:
It's not so much about generating secure DDS artifacts, and more about modeling the permissions that process may have as a whole. Let's say you have two Bridge processes that interface with domains A and C respectively, and communicate with each other over domain B. If I had a particularly sensitive topic in a, how would a design tool be used to check/verify that it's propagation to domain C is impossible. If this was over one domain, the current model of context could be used to associate the subscribe access of the first Bridge node as a potential information leak to subscribers of the second bridge note, given the transitive probation of data from the first bridge flowing into the second bridge.

Reply:
Compromise reached by passing necessary metadata for bridging in policy context.
See e387ac17364ef89e90ff7a9e68d475350347b4bf